python/lsst/summit/extras/plotting/focusSweep.py

```diff
@@ -0,0 +1,282 @@
+    Populate focus sweep table.
+
+    Parameters
+    ----------
+    records : list of `DimensionRecord`
+        Records spanning focus sweep.
+    consDbClient: `ConsDbClient`
+        Consolidated database client
+    efdClient: `EfdClient`
+        Engineering facilities database client
+
+    Returns
+    -------
+    table : astropy.table.Table
+        Table containing hexapod sweep motions and quick look PSF measurements.
+    """
+    visitString = ",".join(str(r.id) for r in records)
+    data = consDbClient.query(
+        "SELECT "
+        "visit_id as visit_id, "
+        "n_inputs as n_inputs, "
+        "psf_sigma_median as sigma, "
+        "psf_ixx_median as ixx, "
+        "psf_iyy_median as iyy, "
+        "psf_ixy_median as ixy "
+        f"from cdb_lsstcomcamsim.visit1_quicklook WHERE visit_id in ({visitString}) "
+        "ORDER BY visit_id"
+    )
+    data["seqNum"] = data["visit_id"] % 10000
+    data["T"] = data["ixx"] + data["iyy"]
+    data["e1"] = (data["ixx"] - data["iyy"]) / data["T"]
+    data["e2"] = 2 * data["ixy"] / data["T"]
+    data["fwhm"] = np.sqrt(np.log(256)) * PLATESCALE * data["sigma"]
+
+    # Placeholder columns for EFD data
+    for prefix in ["cam_", "m2_"]:
+        for k in ["x", "y", "z", "u", "v", "age"]:
+            data[prefix + k] = np.nan
+
+    for row in data:
+        rIdx = [r.id for r in records].index(row["visit_id"])
+        record = records[rIdx]
+        for pid, prefix in zip(["MTHexapod:1", "MTHexapod:2"], ["cam_", "m2_"]):
+            try:
+                efdData = getMostRecentRowWithDataBefore(
+                    efdClient,
+                    "lsst.sal.MTHexapod.logevent_compensatedPosition",
+                    timeToLookBefore=record.timespan.begin,
+                    maxSearchNMinutes=3,
+                    where=lambda df: df["private_identity"] == pid,
+                )
+            except ValueError:
+                row[prefix + "age"] = np.nan
+                for k in ["x", "y", "z", "u", "v"]:
+                    row[prefix + k] = np.nan
+            else:
+                age = record.timespan.begin - efdTimestampToAstropy(efdData["private_efdStamp"])
+                row[prefix + "age"] = age.sec
+                for k in ["x", "y", "z", "u", "v"]:
+                    row[prefix + k] = efdData[k]
+    return data
+
+
+def inferSweepVariable(data):
+    """Heuristically determine which variable is being swept during a focus
+    sweep.
+
+    Parameters
+    ----------
+    data : astropy.table.Table
+        Table holding sweep hexapod motions and PSF measurements.
+
+    Returns
+    -------
+    varName : str | None
+        Name of the inferred active hexapod variable or None if inference
+        failed.
+    """
+    # Examine the ratio of RMS hexapod values to RMS hexapod residuals from a
+    # linear fit against seqNum.  If removing the linear term significantly
+    # reduces the RMS, that's a good sign this is the active variable.
```

Comment:
Oof.  I guess you did say "heuristically".

Reply:
Yeah, there are plans for better ways to do this, but extracting things out of the control system which aren't already in the EFD or the headers is pretty tricky, and we needed to get on with things. Definitely a big "oof" though, for sure.